### DISCUSSION TITLE
Latest Docfx version v2.70.0 code-snippet color issue

### DISCUSSION THREAD

**Discussion by kavinkumarrajendran2000** [dotnet/docfx#9145] [General]
Hi Everyone,

Currently I have been facing an issue(code-snippet color issue and the language detection in code-snippet) while generating a code-snippet in API pages using latest docfx version v2.70.0.

But while using my older docfx version v2.56.1 the code-snippet color is perfect and language detection for code-snippet also works perfectly.

**using older version(v2.56.1) output:**
<img width="389" alt="image" src="https://github.com/dotnet/docfx/assets/97512284/88d1465f-8408-406a-ba35-4ff7d80ef9c9">

  **using latest version(v2.70.0) output:**  
<img width="373" alt="image" src="https://github.com/dotnet/docfx/assets/97512284/bbd822a9-2c5d-45e9-95e8-396332230996">

Is this the issue or it was deprecated in the latest version(v2.70.0)?

How to achieve the older(v2.56.1) code-snippet color effect using my latest docfx version(v2.70.0) if it was deprecated? Could you please help me to overcome this issue.

Regards,
Kavin

**Comment by filzrev** [comment]
It seems `auto-detected language by highlight.js` is something wrong value selected.
What's language is selected on your environment?
It can be confirmed by `code` element's CSS class. (e.g. `<code class="lang-xml`)

If language is not selected properly.
It need to create custom template. to customize `Highlight.js` settings.

**How to customize template**
https://dotnet.github.io/docfx/docs/template.html?tabs=modern#custom-template

**Example of `public/main.js`**
```js
export default {
    configureHljs: (hljs) => {
        // Customize [hightlight.js](https://highlightjs.readthedocs.io/en/latest/api.html#configure) options
        hljs.configure({ languages: ["xml", "csharp"] })
    }
}
```

**Reply by kavinkumarrajendran2000** [reply]
Hi Filzrev,

Thanks for the update.

The language has been generated as "csharp" even my code lang is XAML.

I have used the above main.js code in my default template of docfx v2.70.0, but still I got the below output with lang "csharp".
<img width="770" alt="image" src="https://github.com/dotnet/docfx/assets/97512284/6facada2-2f8d-4cfe-b2c8-2a53871fd713">

Could you please help me to overcome this issue?

Regards,
Kavin

**Reply by filzrev** [reply]
Is it able to provide reproducible sample code?
I've tested with following steps. but can't reproduce the behaviors. (Xaml content is resolved to `language-csharp`)

## Environment
- docfx 2.70.0
- OS: Windows
- Browser: Chrome

## Steps

1. Create docfx project with `docfx init -q`
2. Edit index.md with following contents
3. Run `docfx build --serve` & confirm rendered HTML.

## Tested Markdown contents

````md
# This is the **HOMEPAGE**.
## [Xaml](#tab/xaml)

```
<!-- Eliminated for simplicity -->
<chart:AreaSeries ItemsSource="{Binding Data}"
                  XBindingPath="XValue"
                  YBindingPath="YValue">
    <chart:AreaSeries.DataLabelSettings>
        <chart:CartesianDataLabelSettings />
    </chart:AreaSeries.DataLabelSettings>
</chart:AreaSeries>
```

## [.NET](#tab/csharp)

```
internal class Program
{
    internal static void Main(string[] args)
    {
    }
}
```

---
````

## Results

1. By default xaml content is resolved to `language-dust`.
2. When configuring languages with `public/main.js`.  it is resolved to `language-xml`.

## Other
I've tested with following [Code Snippet](https://dotnet.github.io/docfx/docs/markdown.html?tabs=linux%2Cdotnet#code-snippet) syntax also.
But results are same as above.

```
## [Xaml](#tab/xaml)

[!code[](Sample.xaml)]

## [.NET](#tab/csharp)

[!code[](Sample.cs)]
```

**Reply by kavinkumarrajendran2000** [reply]
Hi Filzrev,

Please find the xml used for generating the below output.

```
 <example>
             # [Xaml](#tab/tabid-1)
             <code><![CDATA[
                 <chart:SfCartesianChart>
            
                 <!-- ... Eliminated for simplicity-->
            
                      <chart:AreaSeries ItemsSource="{Binding Data}"
                                        XBindingPath="XValue"
                                        YBindingPath="YValue">
                          <chart:AreaSeries.DataLabelSettings>
                              <chart:CartesianDataLabelSettings />
                          </chart:AreaSeries.DataLabelSettings>
                      </chart:AreaSeries>
            
                 </chart:SfCartesianChart>
             ]]>
             </code>
             # [C#](#tab/tabid-2)
             <code><![CDATA[
                 SfCartesianChart chart = new SfCartesianChart();
                 ViewModel viewModel = new ViewModel();
            
                 // Eliminated for simplicity
            
                 AreaSeries series = new AreaSeries();
                 series.ItemsSource = viewModel.Data;
                 series.XBindingPath = "XValue";
                 series.YBindingPath = "YValue";
                 series.DataLabelSettings = new CartesianDataLabelSettings();
                 chart.Series.Add(series);
            
             ]]>
             </code>
             ***
             </example>
``` 


output:

![image](https://github.com/dotnet/docfx/assets/97512284/b908c678-3c77-48c4-988f-52bab4869ac7)

How the language is being generated as lang-csharp?

**Reply by KalleOlaviNiemitalo** [reply]
Is the `example` element part of an XML documentation comment in a C# source file, or is it in a Markdown file?

Possibly relevant "csharp" default from <https://github.com/dotnet/docfx/pull/8607>: <https://github.com/dotnet/docfx/blob/78b4679954c0d9c73488fb118d3e07abba3717da/src/Docfx.Dotnet/Parsers/XmlComment.cs#L145>

**Reply by filzrev** [reply]
I can reproduce the problem by setting supplied text to doc comment.
As noted above comment. Latest docfx explicitly set `lang-csharp` when generating metadata from XML comment.

Currently it seems there is no way to avoid this behaviors.
I've created issue #9156 on this issue.

**Comment by filzrev** [comment]
Latest docfx (`v2.70.2`) include features to specify code element language. (#9156)

Try to explicitly specify following language attribute on doc comments that contains xaml code.
`<code language="xaml">` or `<code language="xml">`

When setting language attribute as `xaml`.
It need to specify `language alias` (`xaml` -> `xml`) in `public/main.js`.
(Because  [`highlight.js` don't have build-in support for xaml](https://github.com/highlightjs/highlight.js/blob/main/SUPPORTED_LANGUAGES.md)).

```js
export default {
  configureHljs: function (hljs) {
    // Customize [hightlight.js](https://highlightjs.readthedocs.io/en/latest/api.html#configure) options
    hljs.registerAliases('xaml', { languageName: 'xml' });
  }
}
```

**Reply by kavinkumarrajendran2000** [reply]
Hi @filzrev ,

I have used the above suggested function in main.js file but still my code-snippet is generated as "lang-csharp" for XML code in docfx version 2.70.0.

**xml document:** 
<img width="414" alt="image" src="https://github.com/dotnet/docfx/assets/97512284/58811611-698b-4ea0-8554-b6d9e85294ec">


**Output:**
<img width="668" alt="image" src="https://github.com/dotnet/docfx/assets/97512284/569ea40a-2a9c-44f2-ad08-e6c961a638aa">


**Docfx.json:**
{
  "metadata": [
    {
      "src": [
        {
          "files": [
            "src/**/*.dll"
          ]
        }
      ],
      "dest": "api",
      "includePrivateMembers": false,
      "disableGitFeatures": false,
      "disableDefaultFilter": false,
      "noRestore": false,
      "namespaceLayout": "flattened",
      "memberLayout": "samePage",
      "allowCompilationErrors": false
    }
  ],
  "build": {
    "content": [
      {
        "files": [
          "api/**.yml",
          "api/index.md"
        ]
      },
      {
        "files": [
          "articles/**.md",
          "articles/**/toc.yml",
          "toc.yml",
          "*.md"
        ]
      }
    ],
    "resource": [
      {
        "files": [
          "images/**"
        ]
      }
    ],
    "output": "_site",
    "globalMetadataFiles": [],
    "fileMetadataFiles": [],
    "template": [
      "default"
    ],
    "postProcessors": [],
    "keepFileLink": false,
    "disableGitFeatures": false
  }
}

**templates\default\styles\main.js:**
export default {
  configureHljs: function (hljs) {
    // Customize [hightlight.js](https://highlightjs.readthedocs.io/en/latest/api.html#configure) options
    hljs.registerAliases('xaml', { languageName: 'xml' });
  }
}

How to fix the language detection issue and color issue in code-snippet?

Regards,
Kavin

**Reply by KalleOlaviNiemitalo** [reply]
@kavinkumarrajendran2000, the `public/main.js` addition is to make the highlighting recognize "xaml" as an alias of "xml".  It doesn't fix the entire problem.  You also need to upgrade to DocFX 2.70.2 and put `<code language="xaml">` in the XML documentation comment.

**Reply by filzrev** [reply]
> but still my code-snippet is generated as "lang-csharp" for XML code in docfx version 2.70.0.

Try using latest version of docfx (`v2.70.2`).
If it isn't worked using latest docfx. Please provide minimal reproduceable xml comment by text format.

**Reply by KalleOlaviNiemitalo** [reply]
@filzrev, what do you think would automatically remove the `language` attribute?

In my experience, Roslyn keeps even `xml:lang` attributes (which I use when transforming the compiler-generated XML file to `xs:documentation` elements in an XML schema).

**Reply by filzrev** [reply]
@KalleOlaviNiemitalo 
Sorry it's my misunderstanding. I've removed above comments.

**Reply by kavinkumarrajendran2000** [reply]
@filzrev ,

Thanks for the update, the code-snippet color issue has been fixed in latest version v2.70.2 when using code language="xml" in the XML documentation.

Is there any way to make latest docfx to auto detect the language by customizing the template without mentioning the language="xml" in code tag?

**Reply by KalleOlaviNiemitalo** [reply]
Maybe it could be done with JavaScript; search for code snippets that have language "csharp", check whether the first non-whitespace character is a less-than sign, and flip the language to "xml".

**Reply by filzrev** [reply]
> Is there any way to make latest docfx to auto detect the language by customizing the template without mentioning the language="xml" in code tag?

Currently there is no way to revert old docfx behaviors without source code modification.

If there is specific reason that require `auto language detection by highlighs.js `.
It needs to create `feature request` issue.

Another option is removing css class by following steps.
1. Run `docfx metadata` command
2. Remove ` class=lang-charp`  class from generated yaml files.
3. Run `docfx build` to build document

**Reply by kavinkumarrajendran2000** [reply]
@filzrev ,

It is working when we remove class=lang-csharp class from generated yaml files. But there are many yaml files, I can't remove manually.

So, I tried to remove using replace action in docfx.json file in two ways, but it is not working. can you suggest how to do it using docfx.json.

**1)Docfx.json**
"globalMetadata": {
            "script": {
                "type": "script",
                "value": "for (var file in files) { if (file.endsWith('.yml')) { var content = files[file].content; content = content.replace(/<code class=\"lang-csharp\">/g, <code>); files[file].content = content; } }"
            }
        }
        
  
<img width="365" alt="image" src="https://github.com/dotnet/docfx/assets/97512284/b71ebf8d-9bfd-4fac-a77c-e796ce0955e8">
